From 4a04e4561c17578542e301187168b0152d0596cf Mon Sep 17 00:00:00 2001
From: Super User <root@localhost.localdomain>
Date: Sat, 9 May 2026 01:48:42 +0800
Subject: [PATCH 5/5] KVM: arm64: HDBSS v4 implementation

- Simple fixes (HDBSSBR_EL2 define, size->order)
- Add dirty_bit.c for HDBSS code
- DBM bug fixes (kvm_pgtable_stage2_pte_prot, stage2_pte_needs_update)
- Auto enable/disable HDBSS during migration
- Enable eager hugepage splitting when HDBSS available
---
 arch/arm64/include/asm/kvm_dirty_bit.h |  42 +++++
 arch/arm64/include/asm/kvm_host.h      |   2 +-
 arch/arm64/kvm/Makefile                |   1 +
 arch/arm64/kvm/arm.c                   |  81 +---------
 arch/arm64/kvm/dirty_bit.c             | 216 +++++++++++++++++++++++++
 arch/arm64/kvm/hyp/pgtable.c           |   5 +
 arch/arm64/kvm/mmu.c                   |   8 +-
 arch/arm64/kvm/virtcca_cvm.c           |  24 +++
 8 files changed, 296 insertions(+), 83 deletions(-)
 create mode 100644 arch/arm64/include/asm/kvm_dirty_bit.h
 create mode 100644 arch/arm64/kvm/dirty_bit.c

diff --git a/arch/arm64/include/asm/kvm_dirty_bit.h b/arch/arm64/include/asm/kvm_dirty_bit.h
new file mode 100644
index 000000000000..1752d1de9c61
--- /dev/null
+++ b/arch/arm64/include/asm/kvm_dirty_bit.h
@@ -0,0 +1,42 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) 2023, 2024 Hisilicon Limited.
+ *
+ * Kernel-based dirty page tracking for KVM.
+ */
+
+#ifndef __ARM64_KVM_DIRTY_BIT_H__
+#define __ARM64_KVM_DIRTY_BIT_H__
+
+#include <linux/types.h>
+#include <asm/sysreg.h>
+
+struct kvm_vcpu;
+struct kvm;
+
+/* HDBSS entry field definitions */
+#define HDBSS_ENTRY_VALID	BIT(0)
+#define HDBSS_ENTRY_IPA		GENMASK_ULL(55, 12)
+
+/* HDBSS request for flushing buffer */
+#define KVM_REQ_FLUSH_HDBSS	KVM_ARCH_REQ(10)
+
+struct vcpu_hdbss_state {
+	phys_addr_t base_phys;
+	u32 order;		/* page order of the buffer */
+	u32 next_index;
+};
+
+/* HDBSS flush operations */
+void kvm_flush_hdbss_buffer(struct kvm_vcpu *vcpu);
+void kvm_arm_vcpu_free_hdbss(struct kvm_vcpu *vcpu);
+
+/* VM-level HDBSS enable/disable */
+int kvm_cap_arm_enable_hdbss(struct kvm *kvm, struct kvm_enable_cap *cap);
+int kvm_arm_enable_hdbss_internal(struct kvm *kvm, unsigned long order);
+int kvm_arm_disable_hdbss_internal(struct kvm *kvm);
+
+/* Sync dirty pages to slot before reporting to userspace */
+void kvm_flush_hdbss_to_memslot(struct kvm *kvm);
+
+#endif /* __ARM64_KVM_DIRTY_BIT_H__ */
\ No newline at end of file
diff --git a/arch/arm64/include/asm/kvm_host.h b/arch/arm64/include/asm/kvm_host.h
index c3ab46ed5f56..5fe79335ba7b 100644
--- a/arch/arm64/include/asm/kvm_host.h
+++ b/arch/arm64/include/asm/kvm_host.h
@@ -521,7 +521,7 @@ extern u64 kvm_nvhe_sym(hyp_cpu_logical_map)[NR_CPUS];
 
 struct vcpu_hdbss_state {
 	phys_addr_t base_phys;
-	u32 size;
+	u32 order;		/* page order of the buffer */
 	u32 next_index;
 };
 
diff --git a/arch/arm64/kvm/Makefile b/arch/arm64/kvm/Makefile
index 4a7ee6542668..408f4805d4c9 100644
--- a/arch/arm64/kvm/Makefile
+++ b/arch/arm64/kvm/Makefile
@@ -16,6 +16,7 @@ kvm-y += arm.o mmu.o mmio.o psci.o hypercalls.o \
 	 guest.o debug.o reset.o sys_regs.o stacktrace.o \
 	 vgic-sys-reg-v3.o fpsimd.o pkvm.o \
 	 arch_timer.o trng.o vmid.o emulate-nested.o nested.o \
+	 dirty_bit.o \
 	 vgic/vgic.o vgic/vgic-init.o \
 	 vgic/vgic-irqfd.o vgic/vgic-v2.o \
 	 vgic/vgic-v3.o vgic/vgic-v4.o \
diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index 1e834706a87d..fe17f4fab032 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -129,86 +129,7 @@ int kvm_arch_vcpu_should_kick(struct kvm_vcpu *vcpu)
 	return kvm_vcpu_exiting_guest_mode(vcpu) == IN_GUEST_MODE;
 }
 
-void kvm_arm_vcpu_free_hdbss(struct kvm_vcpu *vcpu)
-{
-	struct page *hdbss_pg;
-
-	hdbss_pg = phys_to_page(vcpu->arch.hdbss.base_phys);
-	if (hdbss_pg)
-		__free_pages(hdbss_pg, vcpu->arch.hdbss.size);
-
-	vcpu->arch.hdbss.size = 0;
-}
-
-static int kvm_cap_arm_enable_hdbss(struct kvm *kvm,
-				    struct kvm_enable_cap *cap)
-{
-	unsigned long i;
-	struct kvm_vcpu *vcpu;
-	struct page *hdbss_pg = NULL;
-	__u64 order = cap->args[0];
-	bool enable = cap->args[1] ? true : false;
-
-	if (!system_supports_hdbss())
-		return -EINVAL;
-
-	if (order > HDBSS_MAX_SIZE)
-		return -EINVAL;
-
-	if (!enable && !kvm->arch.enable_hdbss) /* Already Off */
-		return 0;
-
-	if (enable && kvm->arch.enable_hdbss) /* Already On, can't set order */
-		return -EINVAL;
-
-	if (!enable) { /* Turn it off */
-		kvm->arch.mmu.vtcr &= ~(VTCR_EL2_HD | VTCR_EL2_HDBSS | VTCR_EL2_HA);
-
-		kvm_for_each_vcpu(i, vcpu, kvm) {
-			/* Kick vcpus to flush hdbss buffer. */
-			kvm_vcpu_kick(vcpu);
-
-			kvm_arm_vcpu_free_hdbss(vcpu);
-		}
-
-		kvm->arch.enable_hdbss = false;
-
-		return 0;
-	}
-
-	/* Turn it on */
-	kvm_for_each_vcpu(i, vcpu, kvm) {
-		hdbss_pg = alloc_pages(GFP_KERNEL_ACCOUNT, order);
-		if (!hdbss_pg)
-			goto error_alloc;
-
-		vcpu->arch.hdbss = (struct vcpu_hdbss_state) {
-			.base_phys = page_to_phys(hdbss_pg),
-			.size = order,
-			.next_index = 0,
-		};
-	}
-
-	kvm->arch.enable_hdbss = true;
-	kvm->arch.mmu.vtcr |= VTCR_EL2_HD | VTCR_EL2_HDBSS | VTCR_EL2_HA;
-
-	/*
-	 * We should kick vcpus out of guest mode here to load new
-	 * vtcr value to vtcr_el2 register when re-enter guest mode.
-	 */
-	kvm_for_each_vcpu(i, vcpu, kvm)
-		kvm_vcpu_kick(vcpu);
-
-	return 0;
-
-error_alloc:
-	kvm_for_each_vcpu(i, vcpu, kvm) {
-		if (vcpu->arch.hdbss.base_phys)
-			kvm_arm_vcpu_free_hdbss(vcpu);
-	}
-
-	return -ENOMEM;
-}
+#include <asm/kvm_dirty_bit.h>
 
 int kvm_vm_ioctl_enable_cap(struct kvm *kvm,
 			    struct kvm_enable_cap *cap)
diff --git a/arch/arm64/kvm/dirty_bit.c b/arch/arm64/kvm/dirty_bit.c
new file mode 100644
index 000000000000..a502be7c91ec
--- /dev/null
+++ b/arch/arm64/kvm/dirty_bit.c
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2023, 2024 Hisilicon Limited.
+ *
+ * Kernel-based dirty page tracking for KVM.
+ * Contains HDBSS (Hardware Dirty Bit System) support.
+ */
+
+#include <linux/kvm_host.h>
+#include <linux/page-flags.h>
+#include <asm/kvm_dirty_bit.h>
+#include <asm/kvm_mmu.h>
+#include <asm/kvm_host.h>
+#include <asm/esr.h>
+
+void kvm_arm_vcpu_free_hdbss(struct kvm_vcpu *vcpu)
+{
+	struct page *hdbss_pg;
+
+	hdbss_pg = phys_to_page(vcpu->arch.hdbss.base_phys);
+	if (hdbss_pg)
+		__free_pages(hdbss_pg, vcpu->arch.hdbss.order);
+
+	vcpu->arch.hdbss.order = 0;
+}
+
+void kvm_flush_hdbss_buffer(struct kvm_vcpu *vcpu)
+{
+	int idx, curr_idx;
+	u64 br_el2;
+	u64 *hdbss_buf;
+	struct kvm *kvm = vcpu->kvm;
+
+	if (!kvm->arch.enable_hdbss)
+		return;
+
+	curr_idx = HDBSSPROD_IDX(read_sysreg_s(SYS_HDBSSPROD_EL2));
+	br_el2 = HDBSSBR_EL2(vcpu->arch.hdbss.base_phys, vcpu->arch.hdbss.order);
+
+	/* Do nothing if HDBSS buffer is empty or br_el2 is NULL */
+	if (curr_idx == 0 || br_el2 == 0)
+		return;
+
+	hdbss_buf = page_address(phys_to_page(vcpu->arch.hdbss.base_phys));
+	if (!hdbss_buf)
+		return;
+
+	guard(write_lock_irqsave)(&vcpu->kvm->mmu_lock);
+	for (idx = 0; idx < curr_idx; idx++) {
+		u64 gpa;
+
+		gpa = hdbss_buf[idx];
+		if (!(gpa & HDBSS_ENTRY_VALID))
+			continue;
+
+		gpa &= HDBSS_ENTRY_IPA;
+		kvm_vcpu_mark_page_dirty(vcpu, gpa >> PAGE_SHIFT);
+	}
+
+	/* reset HDBSS index */
+	write_sysreg_s(0, SYS_HDBSSPROD_EL2);
+	vcpu->arch.hdbss.next_index = 0;
+	isb();
+}
+
+static int kvm_handle_hdbss_fault(struct kvm_vcpu *vcpu)
+{
+	u64 prod;
+	u64 fsc;
+
+	prod = read_sysreg_s(SYS_HDBSSPROD_EL2);
+	fsc = FIELD_GET(HDBSSPROD_EL2_FSC_MASK, prod);
+
+	switch (fsc) {
+	case HDBSSPROD_EL2_FSC_OK:
+		/* Buffer full, which is reported as permission fault. */
+		kvm_flush_hdbss_buffer(vcpu);
+		return 1;
+	case HDBSSPROD_EL2_FSC_ExternalAbort:
+	case HDBSSPROD_EL2_FSC_GPF:
+		return -EFAULT;
+	default:
+		/* Unknown fault. */
+		WARN_ONCE(1,
+			"Unexpected HDBSS fault type, FSC: 0x%llx (prod=0x%llx, vcpu=%d)\n",
+			fsc, prod, vcpu->vcpu_id);
+		return -EFAULT;
+	}
+}
+
+/*
+ * kvm_arm_enable_hdbss_internal - Enable HDBSS for a VM with given order
+ *
+ * This is called internally during migration initialization.
+ * Returns 0 on success, negative errno on error.
+ */
+int kvm_arm_enable_hdbss_internal(struct kvm *kvm, unsigned long order)
+{
+	unsigned long i;
+	struct kvm_vcpu *vcpu;
+	struct page *hdbss_pg = NULL;
+
+	if (!system_supports_hdbss())
+		return -EINVAL;
+
+	if (order > HDBSS_MAX_SIZE)
+		return -EINVAL;
+
+	if (kvm->arch.enable_hdbss) /* Already On, can't set order */
+		return -EINVAL;
+
+	/* Turn it on */
+	kvm_for_each_vcpu(i, vcpu, kvm) {
+		hdbss_pg = alloc_pages(GFP_KERNEL_ACCOUNT, order);
+		if (!hdbss_pg)
+			goto error_alloc;
+
+		vcpu->arch.hdbss = (struct vcpu_hdbss_state) {
+			.base_phys = page_to_phys(hdbss_pg),
+			.order = order,
+			.next_index = 0,
+		};
+	}
+
+	kvm->arch.enable_hdbss = true;
+	kvm->arch.mmu.vtcr |= VTCR_EL2_HD | VTCR_EL2_HDBSS | VTCR_EL2_HA;
+
+	/*
+	 * We should kick vcpus out of guest mode here to load new
+	 * vtcr value to vtcr_el2 register when re-enter guest mode.
+	 */
+	kvm_for_each_vcpu(i, vcpu, kvm)
+		kvm_vcpu_kick(vcpu);
+
+	return 0;
+
+error_alloc:
+	kvm_for_each_vcpu(i, vcpu, kvm) {
+		if (vcpu->arch.hdbss.base_phys)
+			kvm_arm_vcpu_free_hdbss(vcpu);
+	}
+
+	return -ENOMEM;
+}
+
+/*
+ * kvm_arm_disable_hdbss_internal - Disable HDBSS for a VM
+ *
+ * This is called internally during migration completion.
+ * Returns 0 on success, negative errno on error.
+ */
+int kvm_arm_disable_hdbss_internal(struct kvm *kvm)
+{
+	unsigned long i;
+	struct kvm_vcpu *vcpu;
+
+	if (!kvm->arch.enable_hdbss) /* Already Off */
+		return 0;
+
+	kvm->arch.mmu.vtcr &= ~(VTCR_EL2_HD | VTCR_EL2_HDBSS | VTCR_EL2_HA);
+
+	kvm_for_each_vcpu(i, vcpu, kvm) {
+		/* Kick vcpus to flush hdbss buffer. */
+		kvm_vcpu_kick(vcpu);
+
+		kvm_arm_vcpu_free_hdbss(vcpu);
+	}
+
+	kvm->arch.enable_hdbss = false;
+
+	return 0;
+}
+
+/*
+ * kvm_cap_arm_enable_hdbss - Enable/disable HDBSS for a VM
+ *
+ * This is called from kvm_vm_ioctl_enable_cap() with KVM_CAP_ARM_HW_DIRTY_STATE_TRACK.
+ * For auto-enable during migration, use kvm_arm_enable_hdbss_internal() directly.
+ */
+int kvm_cap_arm_enable_hdbss(struct kvm *kvm, struct kvm_enable_cap *cap)
+{
+	__u64 order = cap->args[0];
+	bool enable = cap->args[1] ? true : false;
+
+	if (enable)
+		return kvm_arm_enable_hdbss_internal(kvm, order);
+	else
+		return kvm_arm_disable_hdbss_internal(kvm);
+}
+
+/*
+ * kvm_flush_hdbss_to_memslot - Flush all CPUs' dirty log buffers to the dirty_bitmap.
+ *
+ * Called before reporting dirty_bitmap to userspace. Send a request with
+ * KVM_REQUEST_WAIT to flush buffer synchronously.
+ */
+void kvm_flush_hdbss_to_memslot(struct kvm *kvm)
+{
+	struct kvm_vcpu *vcpu;
+
+	if (!kvm->arch.enable_hdbss)
+		return;
+
+	kvm_make_all_cpus_request(kvm, KVM_REQ_FLUSH_HDBSS);
+}
+
+/*
+ * kvm_handle_hdbss_trap - Handle HDBSS trap from guest
+ *
+ * Called from kvm_handle_guest_abort when an HDBSS fault is detected.
+ * Returns 1 if the fault was handled, negative errno on error.
+ */
+int kvm_handle_hdbss_trap(struct kvm_vcpu *vcpu)
+{
+	return kvm_handle_hdbss_fault(vcpu);
+}
\ No newline at end of file
diff --git a/arch/arm64/kvm/hyp/pgtable.c b/arch/arm64/kvm/hyp/pgtable.c
index a8646214e2fc..68fd47a6f111 100644
--- a/arch/arm64/kvm/hyp/pgtable.c
+++ b/arch/arm64/kvm/hyp/pgtable.c
@@ -729,6 +729,8 @@ enum kvm_pgtable_prot kvm_pgtable_stage2_pte_prot(kvm_pte_t pte)
 		prot |= KVM_PGTABLE_PROT_W;
 	if (!(pte & KVM_PTE_LEAF_ATTR_HI_S2_XN))
 		prot |= KVM_PGTABLE_PROT_X;
+	if (pte & KVM_PTE_LEAF_ATTR_HI_S2_DBM)
+		prot |= KVM_PGTABLE_PROT_DBM;
 
 	return prot;
 }
@@ -738,6 +740,9 @@ static bool stage2_pte_needs_update(kvm_pte_t old, kvm_pte_t new)
 	if (!kvm_pte_valid(old) || !kvm_pte_valid(new))
 		return true;
 
+	if ((old ^ new) & KVM_PTE_LEAF_ATTR_HI_S2_DBM)
+		return true;
+
 	return ((old ^ new) & (~KVM_PTE_LEAF_ATTR_S2_PERMS));
 }
 
diff --git a/arch/arm64/kvm/mmu.c b/arch/arm64/kvm/mmu.c
index 4daa11ad4dd4..18ef806c8dc3 100644
--- a/arch/arm64/kvm/mmu.c
+++ b/arch/arm64/kvm/mmu.c
@@ -941,8 +941,12 @@ int kvm_init_stage2_mmu(struct kvm *kvm, struct kvm_s2_mmu *mmu, unsigned long t
 	for_each_possible_cpu(cpu)
 		*per_cpu_ptr(mmu->last_vcpu_ran, cpu) = -1;
 
-	 /* The eager page splitting is disabled by default */
-	mmu->split_page_chunk_size = KVM_ARM_EAGER_SPLIT_CHUNK_SIZE_DEFAULT;
+	 /* The eager page splitting is disabled by default if system has no HDBSS */
+	if (system_supports_hdbss())
+		mmu->split_page_chunk_size = PAGE_SIZE;
+	else
+		mmu->split_page_chunk_size = KVM_ARM_EAGER_SPLIT_CHUNK_SIZE_DEFAULT;
+
 	mmu->split_page_cache.gfp_zero = __GFP_ZERO;
 
 	mmu->pgt = pgt;
diff --git a/arch/arm64/kvm/virtcca_cvm.c b/arch/arm64/kvm/virtcca_cvm.c
index 44283d2284d7..a0646d27221e 100644
--- a/arch/arm64/kvm/virtcca_cvm.c
+++ b/arch/arm64/kvm/virtcca_cvm.c
@@ -15,6 +15,7 @@
 #include <asm/kvm_mmu.h>
 #include <asm/virtcca_coda.h>
 #include <asm/kvm_tmm.h>
+#include <asm/kvm_dirty_bit.h>
 #include <asm/stage2_pgtable.h>
 #include <asm/virtcca_cvm_host.h>
 #include <asm/virtcca_coda.h>
@@ -848,6 +849,16 @@ static int kvm_populate_ipa_cvm_range(struct kvm *kvm,
 		if (cvm->mig_state->mig_src == VIRTCCA_MIG_DST) {
 			cvm->mig_state->mig_src = VIRTCCA_MIG_SRC;
 			kvm_info("the ipa range is populated before migraion\n");
+
+			/*
+			 * Auto-enable HDBSS for destination VM during migration.
+			 * Use default order (HDBSS_MAX_SIZE = 9 = 2MB buffer per vcpu).
+			 * This is called during memory population which is part of
+			 * migration initialization.
+			 */
+			if (system_supports_hdbss())
+				kvm_arm_enable_hdbss_internal(kvm, HDBSS_MAX_SIZE);
+
 			return 0;
 		}
 	}
@@ -2646,6 +2657,13 @@ static int virtcca_mig_import_end(struct kvm *kvm)
 	cvm->swiotlb_start = 0;
 	cvm->swiotlb_end = 0;
 
+	/*
+	 * Auto-disable HDBSS when migration completes (import_end).
+	 * This is called after the migration data has been imported.
+	 */
+	if (kvm->arch.enable_hdbss)
+		kvm_arm_disable_hdbss_internal(kvm);
+
 	WRITE_ONCE(cvm->state, CVM_STATE_ACTIVE);
 
 	return 0;
@@ -3505,6 +3523,12 @@ static int virtcca_mig_export_abort(struct kvm *kvm)
 	cvm->swiotlb_start = 0;
 	cvm->swiotlb_end = 0;
 
+	/*
+	 * Auto-disable HDBSS when migration aborts.
+	 */
+	if (kvm->arch.enable_hdbss)
+		kvm_arm_disable_hdbss_internal(kvm);
+
 	return ret;
 }
 
-- 
2.43.0

